Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -116,271 +53,113 @@ COREGRAPHICS_EXPORT void CGImageAddDestructionListener(CGImageDestructionListene
+ @Status Interoperable
 */
-CGImageRef CGImageCreateWithPNGDataProvider(CGDataProviderRef source,
-                                            const CGFloat decode[],
-                                            bool shouldInterpolate,
-                                            CGColorRenderingIntent intent) {
-    FAIL_FAST_IF_FALSE(decode == nullptr);
-    FAIL_FAST_HR_IF_FALSE_MSG(E_INVALIDARG,
-                              ((source == nullptr) || [(NSObject*)source isKindOfClass:[NSData class]]),
-                              "CGDataProviderRef does not derive from NSData!");
-
-    id img = [[_LazyUIImage alloc] initWithData:(NSData*)source];
-
-    return (CGImageRef)[img CGImage];
+CFTypeID CGImageGetTypeID() {
+    return __CGImage::GetTypeID();
 }
 
 /**
  @Status Caveat
- @Notes Doesn't support copy-on-write semantics - returns an unlinked copy of the source
-        image cropped to the specified rectangle.
+ @Notes Limited bitmap formats available and decode is not supported.
 */
-CGImageRef CGImageCreateWithImageInRect(CGImageRef ref, CGRect rect) {
-    if (ref == NULL) {
-        TraceWarning(TAG, L"CGImageCreateWithImageInRect: ref = NULL!");
-        return 0;
-    }
-
-    rect = CGRectIntegral(rect);
+CGImageRef CGImageCreate(size_t width,
+                         size_t height,
+                         size_t bitsPerComponent,
+                         size_t bitsPerPixel,
+                         size_t bytesPerRow,
+                         CGColorSpaceRef colorSpace,
+                         CGBitmapInfo bitmapInfo,
+                         CGDataProviderRef provider,
+                         const float* decode,
+                         bool shouldInterpolate,
+                         CGColorRenderingIntent intent) {
+    RETURN_NULL_IF(((provider == nullptr) || ![(NSObject*)provider isKindOfClass:[NSData class]]));
 
-    CGRect imgRefSize;
-    imgRefSize.origin.x = 0;
-    imgRefSize.origin.y = 0;
-    imgRefSize.size.width = (float)ref->Backing()->Width();
-    imgRefSize.size.height = (float)ref->Backing()->Height();
+    NSData* dataProvider = (__bridge NSData*)provider;
 
-    rect = CGRectIntersection(rect, imgRefSize);
+    unsigned char* data = (unsigned char*)[dataProvider bytes];
 
-    __CGSurfaceInfo surfaceInfo;
-    ref->Backing()->GetSurfaceInfoWithoutPixelPtr(&surfaceInfo);
+    bool colorSpaceAllocated = false;
+    if (colorSpace == NULL) {
+        if (bytesPerRow >= (width * 3)) {
+            TraceWarning(TAG, L"Warning: colorSpace = NULL, assuming RGB based on bytesPerRow.");
+            colorSpace = CGColorSpaceCreateDeviceRGB();
+        } else {
+            TraceWarning(TAG, L"Warning: colorSpace = NULL, assuming Gray based on bytesPerRow.");
+            colorSpace = CGColorSpaceCreateDeviceGray();
+        }
 
-    // Override width and height with the rect
-    surfaceInfo.width = rect.size.width;
-    surfaceInfo.height = rect.size.height;
+        colorSpaceAllocated = true;
```

Comment:
Reading more it looks like you are actually just stomping the function argument here to use it as a local variable that may or may not be passed in. In general this is another bad pattern as it makes future code maintainers potentially unsure of the preconditions and the validity of arguments at all points throughout the function. I'd prefer us to write clean and clear code and let the compiler do its magic to save a local if that is what you are concerned about. 

To do this sort of thing in an RAII friendly pattern, you would use some construct that does a +1 on assignment (similar to say shared_ptr instead of unique_ptr in the unique_cf example) and then assign from the argument or .attach if you create locally. That way the RAII object is always at the correct refcount and the release happens at the end without harm.

---
In reply to: [84754523](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84754523) [](ancestors = 84754523)